File: packages/b2c-tooling-sdk/src/operations/code/scripts-types.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Canonical types and backend interface for code-version (Scripts) operations.
 *
 * The OCAPI Data API and the SCAPI Scripts API both manage code versions on a
 * B2C instance. We expose a single canonical shape here so command code is
 * agnostic to which backend serves the request.
 *
 * @module operations/code/scripts-types
 */
import type {BackendBase} from '../../clients/scapi-backend-utils.js';

/**
 * Canonical code version. CamelCase fields match SCAPI; OCAPI mapping
 * converts from snake_case.
 */
export interface CodeVersionInfo {
  id: string;
  active?: boolean;
  cartridges?: string[];
  compatibilityMode?: string;
  activationTime?: string;
  lastModificationTime?: string;
  rollback?: boolean;
  totalSize?: number;
  webDavUrl?: string;
  /** Original backend response, for advanced consumers. */
  _raw?: unknown;
}

/**
 * Backend contract for code-version operations.
 *
 * Reload is implemented as a backend-agnostic helper (`reloadCodeVersion`
 * in the operations module) since it's just `activate(alternate) +
 * activate(target)` on top of these primitives.
 */
export interface ScriptsBackend extends BackendBase {
  listCodeVersions(): Promise<CodeVersionInfo[]>;
  getActiveCodeVersion(): Promise<CodeVersionInfo | undefined>;
  activateCodeVersion(codeVersionId: string): Promise<void>;
  deleteCodeVersion(codeVersionId: string): Promise<void>;
  createCodeVersion(codeVersionId: string): Promise<void>;
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/operations/code/scripts_types.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Backend-agnostic code-version types shared by the OCAPI and SCAPI script backends.

Mirrors ``src/operations/code/scripts-types.ts``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol, runtime_checkable

from b2c_tooling_sdk.clients.scapi_backend_utils import BackendBase
from b2c_tooling_sdk.operations.code.versions import CodeVersionActivationResult


@dataclass
class CodeVersionInfo:
    """A code version, normalized across the OCAPI and SCAPI representations.

    ``activation_time``/``last_modification_time`` are kept as ISO-8601 strings
    (matching the TypeScript ``string`` fields) even though the underlying
    generated Pydantic models parse them as ``datetime`` — each backend
    converts via ``.isoformat()`` at the mapping boundary.
    """

    id: str
    active: bool | None = None
    cartridges: list[str] | None = None
    compatibility_mode: str | None = None
    activation_time: str | None = None
    last_modification_time: str | None = None
    rollback: bool | None = None
    total_size: int | None = None
    web_dav_url: str | None = None
    #: The raw backend-specific payload, for callers that need details this
    #: normalized shape doesn't expose.
    raw: Any = None


@runtime_checkable
class ScriptsBackend(BackendBase, Protocol):
    """Backend-agnostic interface for code-version operations.

    Implemented by :class:`~b2c_tooling_sdk.operations.code.ocapi_scripts_backend.OcapiScriptsBackend`
    and :class:`~b2c_tooling_sdk.operations.code.scapi_scripts_backend.ScapiScriptsBackend`.
    """

    async def list_code_versions(self) -> list[CodeVersionInfo]: ...

    async def get_active_code_version(self) -> CodeVersionInfo | None: ...

    async def activate_code_version(self, code_version_id: str) -> CodeVersionActivationResult: ...

    async def delete_code_version(self, code_version_id: str) -> None: ...

    async def create_code_version(self, code_version_id: str) -> None: ...


__all__ = ["CodeVersionInfo", "ScriptsBackend"]
